src/runtime/qurt_hvx.cpp

```diff
@@ -52,4 +53,214 @@ WEAK void halide_qurt_hvx_unlock_as_destructor(void *user_context, void * /*obj*
+    static uint32_t dbg_cnt = 0;  // limit how many calls debug is generated
+    dbg_cnt++;
+    DBG_PREFETCH("halide_hexagon_prefetch_buffer_t(%u, %d)\n", boxdim, elem_size, num_elem);
+
+    DBG_PREFETCH("  addr range 0x%p => 0x%p\n", addr_beg, addr_end);
+
+    DBG_PREFETCH("  buf host=0x%p elem_size=%d\n", addr, buf->elem_size);
+    for (uint32_t i = 0; i < boxdim; i++) {
+        DBG_PREFETCH("  buf stride[%d]=0x%-8x min[%d]=%-6d ext[%d]=%d\n",
+                          i, stride[i], i, min[i], i, extent[i]);
+    }
+#endif
+
+    // Compute starting position of box
+    int32_t startpos = 0;
+    for (uint32_t i = 0; i < boxdim; i++) {
+        startpos += min[i] * stride[i];
+    }
+    addr += startpos * elem_size;
+    DBG_PREFETCH("  +startpos=0x%x*%d => addr=0x%p\n", startpos, elem_size, addr);
+
+    // Range check starting address
+    if ((addr < addr_beg) || (addr >= addr_end)) {
+        DBG_PREFETCH_V("  l2fetch: 0x%p out of range [0x%p, 0x%p]\n",
+                        addr, addr_beg, addr_end);
+        return 0;
+    }
+
+    // Compute 2-D prefetch descriptor
+    // l2fetch(Rs,Rtt): 48 bit descriptor
+    uint32_t pdir    = 1;       // Prefetch direction: 0=rows, 1=columns
+    uint32_t pstride = stride[0] * elem_size;
+    uint32_t pwidth  = extent[0] * elem_size;
+    uint32_t pheight = 1;       // Initially 1-D descriptor
+
+    uint32_t iterdim = 1;       // Dimension to iterate over
+    if (boxdim > 1) {           // Potential for 2-D descriptor
+        uint32_t newpstride = stride[1] * elem_size;
+        if (newpstride == (newpstride & MASK16)) {
+            pstride = newpstride;
+            pheight = extent[1];
+            iterdim++;
+        }
+    }
+    DBG_PREFETCH("  iterdim:%d\n", iterdim);
+
+#if 0   // TODO: Opt: Box collapse disabled for now - needs more testing, and
+        // TODO       collapse candidates tend to exceed the stride mask size
+    // For boxes with dimension > 2 try to "collapse" any unit height
+    // dimensions by increasing the descriptor stride
+    uint32_t newdim = boxdim;
+    while (newdim > 2) {
+        uint32_t newpstride = 0;
+        if (pheight == 1) {     // if height is currently 1
+            newpstride = stride[iterdim-1] * elem_size;  // update stride
+        } else {
+            break;              // otherwise, we're done collapsing
+        }
+        if (newpstride == (newpstride & MASK16)) {  // and if it fits in mask...
+            pstride = newpstride;             // ...accept new stride
+            pheight = extent[iterdim];        // ...and height
+        } else {
+            break;              // otherwise, we're done collapsing
+        }
+        newdim--;       // remaining non-collapsed dimensions
+        iterdim++;      // update innermost iterate dimension
+    }
+    DBG_PREFETCH("  iterdim:%d\n", iterdim);
+#endif
+
+    pdir    = pdir & 0x1;           // bit  48
+    pstride = MIN(pstride, MASK16); // bits 47:32
```

Comment:
The min is there just to keep values within the mask (and to saturate, not just mask the bits that fit).  If pstride didn't fit, the check above would have kept the prefetch descriptor to a single row and the 2nd dimension would be used for iterating over 1-D descriptors.  If width or height doesn't fit in MASK16, then we only get a partial prefetch....but still a correct one.